numba/targets/arraymath.py

```diff
@@ -1380,6 +1380,81 @@ def np_vander_seq_impl(x, N=None, increasing=False):
     elif isinstance(x, (types.Tuple, types.Sequence)):
         return np_vander_seq_impl
 
+#----------------------------------------------------------------------------
+# Mathematical functions
+
+@register_jitable
+def np_interp_impl_inner(x, xp, fp, out_dtype):
+    x_arr = np.asarray(x)
+    xp_arr = np.asarray(xp)
+    fp_arr = np.asarray(fp)
+
+    if len(xp_arr) == 0:
+        raise ValueError('array of sample points is empty')
+
+    if len(xp_arr) != len(fp_arr):
+        raise ValueError('fp and xp are not of the same length.')
+
+    if xp_arr.size > 1 and not np.all(np.diff(xp_arr) > 0):
+        raise ValueError('The x-coordinates of the data points must be increasing')
+        # this is a strict and literal interpretation of NumPy docs:
+        # https://github.com/numpy/numpy/blob/v1.15.0/numpy/lib/function_base.py#L1192
+        #   "The x-coordinates of the data points, must be increasing if argument
+        #   `period` is not specified."
+        # NumPy does not check for this or enforce it; whereas NumPy will return output,
+        # this implementation will raise.
+
+    if xp_arr.size == 1:
+        return np.full(x_arr.shape, fill_value=fp_arr[0], dtype=out_dtype)
+    else:
+        out = np.empty(x_arr.shape, dtype=out_dtype)
+
+        for i in range(x_arr.size):
+            if x_arr.flat[i] > xp_arr[-1]:
+                out.flat[i] = fp_arr[-1]
+            elif x_arr.flat[i] < xp_arr[0]:
+                out.flat[i] = fp_arr[0]
+            else:
+                idx = np.searchsorted(xp_arr, x_arr.flat[i])
+                f = (x_arr.flat[i] - xp_arr[idx - 1]) / (xp_arr[idx] - xp_arr[idx - 1])
+                out.flat[i] = fp_arr[idx - 1] + f * (fp_arr[idx] - fp_arr[idx - 1])
```

Comment:
This alg ordering is perhaps less conventional, think there's usually a computation of the gradient across the indexed points, then that is multiplied by the `x_arr` fraction and added to the `fp_arr`. The method present has more operations performed in the same variable space (fraction in `x` space), I can think of bad cases for either way!

Reply:
Yeah, fair enough - I guess it depends on the number and distribution of points (x) versus calibration data (xp, fp).  My 'real world' data sets are massively skewed towards the above being less expensive, but I agree that's not a safe generalisation.  Will change it!

Reply:
I think NumPy does a check on the size of `x` and `xp` to decide whether to pre-cache the gradients.  This is imperfect as you'd need to also know the distribution to figure out whether pre-caching is a good idea, but the cost of figuring that out probably outweighs the benefit.

Anyway, I flipped it to pre-cache gradients always.

Reply:
Yeah, it probably either needs a branch or to just lazily populate, `NaN` could be the sentinel for `not computed yet` with the occasion of `NaN` in the inputs causing repeated computations? Or just branch on size of input?